sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+		ocspHosts = conf.OcspServersOverride
+	}
+
+	var wg sync.WaitGroup
+
+	ocspStatuses := make([]*ocspStatus, len(ocspHosts))
+	ocspResponses := make([]*ocsp.Response, len(ocspHosts))
+	errors := make([]error, len(ocspHosts))
+
+	for i, ocspHost := range ocspHosts {
+		u, err := url.Parse(ocspHost)
+		if err != nil {
+			return nil, err
+		}
+
+		hostname := u.Hostname()
+
+		headers := make(map[string]string)
+		headers[httpHeaderContentType] = ocspRequestContentType
+		headers[httpHeaderAccept] = ocspResponseContentType
+		headers[httpHeaderContentLength] = strconv.Itoa(len(ocspReq))
+		headers[httpHeaderHost] = hostname
+		timeout := defaultOCSPResponderTimeout
+
+		ocspClient := retryablehttp.NewClient()
+		ocspClient.HTTPClient.Timeout = timeout
+		ocspClient.HTTPClient.Transport = newInsecureOcspTransport(conf.ExtraCas)
+
+		doRequest := func() error {
+			if conf.QueryAllServers {
+				defer wg.Done()
+			}
+			ocspRes, _, ocspS, err := c.retryOCSP(
+				ctx, ocspClient, retryablehttp.NewRequest, u, headers, ocspReq, issuer)
+			ocspResponses[i] = ocspRes
+			if err != nil {
+				errors[i] = err
+				return err
+			}
+			if ocspS.code != ocspSuccess {
+				ocspStatuses[i] = ocspS
+				return nil
+			}
+
+			ret, err := validateOCSP(ocspRes)
+			if err != nil {
+				errors[i] = err
+				return err
+			}
+			if isValidOCSPStatus(ret.code) {
+				ocspStatuses[i] = ret
+			}
+			return nil
+		}
+		if conf.QueryAllServers {
+			wg.Add(1)
+			go doRequest()
+		} else {
+			err = doRequest()
+			if err == nil {
+				break
+			}
+		}
+	}
+	if conf.QueryAllServers {
+		wg.Wait()
+	}
+	// Good by default
+	var ret *ocspStatus
+	ocspRes := ocspResponses[0]
+	var firstError error
+	for i := range ocspHosts {
+		if errors[i] != nil {
+			if firstError == nil {
+				firstError = errors[i]
+			}
+		} else if ocspStatuses[i] != nil {
+			switch ocspStatuses[i].code {
+			case ocspStatusRevoked:
+				ret = ocspStatuses[i]
+				ocspRes = ocspResponses[i]
+				break
+			case ocspStatusGood:
+				// Use this response only if we don't have a status already, or if what we have was unknown
+				if ret == nil || ret.code == ocspStatusUnknown {
+					ret = ocspStatuses[i]
+					ocspRes = ocspResponses[i]
+				}
+			case ocspStatusUnknown:
+				if ret == nil {
+					// We may want to use this as the overall result
+					ret = ocspStatuses[i]
+					ocspRes = ocspResponses[i]
+				}
+			}
+		}
+	}
+
+	// If no server reported the cert revoked, but we did have an error, report it
+	if (ret == nil || ret.code == ocspStatusUnknown) && firstError != nil {
+		return nil, firstError
+	}
+	// otherwise ret should contain a response for the overall request
+
+	if !isValidOCSPStatus(ret.code) {
+		return ret, nil
+	}
+	v := ocspCachedResponse{
+		time:       float64(time.Now().UTC().Unix()),
+		producedAt: float64(ocspRes.ProducedAt.UTC().Unix()),
+		thisUpdate: float64(ocspRes.ThisUpdate.UTC().Unix()),
+		nextUpdate: float64(ocspRes.NextUpdate.UTC().Unix()),
+	}
```

Comment:
Don't we need to set status here, out of curiosity?

Reply:
I would think so